buckets/redis/bucket_factory.go

```diff
@@ -164,6 +168,67 @@ func (bf *bucketFactory) reconnectToRedis(oldClient *redis.Client) {
 	}
 }
 
+func (bf *bucketFactory) handleConnectionFailure(oldClient *redis.Client) {
+	bf.Lock()
+	defer bf.Unlock()
+
+	if oldClient == bf.client && !bf.connectionNeedsResolution {
+		logging.Print("Attempting to establish new connection to redis")
+		bf.connectionNeedsResolution = true
+		go bf.establishNewConnectionToRedis(oldClient)
+	}
+}
+
+func (bf *bucketFactory) establishNewConnectionToRedis(oldClient *redis.Client) {
+	client := oldClient
+	numsTried := 0
+	exponentialDelay := 1 * time.Second
+	exponentialDelayCeiling := 3 * time.Minute
+
+	for disconnected := true; disconnected; {
```

Comment:
I think there is value in an upper-bound (configurable) of either number of attempts or time spent trying to reconnect before a hard-fail. Otherwise may be cases where your only recourse is to kill the process.